### ISSUE TITLE
Add Adjacency Matrix Implementation.

### ISSUE THREAD

**Issue by vincentk** [vincentk/dedekind#36]

**Comment by vincentk** [comment]
Triage 2026-04-14: the issue body is empty. Could you clarify the intended scope? Specifically: (1) Which graph representation should the adjacency matrix wrap -- std::vector<std::vector<bool>>, a sparse map, or something else? (2) Should it live in the :sequences module, a new :graphs module, or somewhere else? (3) Are there specific category/algebra concepts (e.g. functor, ring) that the adjacency matrix should satisfy? Without this context it is not actionable with good confidence.

**Comment by vincentk** [comment]
I think for an MVP, the following might be nice:
* define a first version of directed graph as a partial order on a suitable set.

Configure a few test cases where e.g. the
* ambient set is some interval on the integers,
* the relation is some simple structure.

Solve some textbook problem such as min-cut/max-flow. Or determine the principal eigenvector after stochastic normalization using the power method. Both would subsume #37 .

**Comment by vincentk** [comment]
It's clear that the word `Matrix` was maybe imprecise here. What I meant was linear map.

**Comment by vincentk** [comment]
Now that IntegerInterval<T> (PR #141) is in the fold, the missing pieces for a linear map are all grounded:

- **Index set**: IntegerInterval<int>{0, n} -- simultaneously a finite set with unique elements, a convex set, and an enumerable terminal set (IsConvexEnumerable).
- **Vector**: Path<double> bounded by size() of an IntegerInterval -- the first n terms form a well-typed finite-dimensional vector.
- **Linear map**: an IsArrow from a bounded Path (domain R^n) to another bounded Path (codomain R^m), parameterised by the two index sets.

The adjacency matrix of a graph is the special case where the map is square (same index set for domain and codomain) and entries are Boolean/integer -- which directly subsumes the MVP sketch above (partial order on an integer interval, stochastic normalization via the power method).

**Comment by vincentk** [comment]
## MVP scope proposal: bona fide linear algebra support

Now that the codebase has the following ingredients,

- `IntegerInterval<T>` as a finite, enumerable, convex index set,
- `Vector<F, N>` as the concrete `N`-dimensional carrier,
- `OneDimensionalVector<S>` as the 1D module/vector-space baseline,
- `dot`, `norm`, `OneForm`, `TwoForm`, and `IsLinearMorphism`,

I think the smallest honest MVP is **not** “all of matrix algebra”, but rather:

### MVP: a concrete finite-dimensional linear map carrier

Introduce a first concrete reification of a linear map

`LinearMap<F, M, N>` or `Matrix<F, M, N>`

with the interpretation

`F^N -> F^M`

and make it work with the existing `Vector<F, N>` carrier.

### Minimal feature set

1. **Concrete carrier**
   - Dense storage for coefficients, likely `std::array<F, M * N>` or `std::array<std::array<F, N>, M>`.
   - `using scalar_type = F;`
   - static row/column extents.

2. **Core operations**
   - matrix-vector application: `A(v)` or `A * v`
   - matrix-matrix composition: `A * B`
   - additive structure: `A + B`, `A - B`
   - scalar multiplication: `s * A`, `A * s`
   - identity map for square case: `identity_linear_map<F, N>()`
   - zero map

3. **Categorical / algebraic grounding**
   - Treat `LinearMap<F, M, N>` as the first concrete witness of `IsLinearMorphism`.
   - For square matrices, check monoid structure under composition.
   - For all shapes, check additive abelian-group-like signatures where available.

4. **Tests that prove the story**
   - `I(v) == v` for identity.
   - `(A * B)(v) == A(B(v))` for composition.
   - linearity: `A(x + y) == A(x) + A(y)` and `A(s * x) == s * A(x)`.
   - small concrete examples in dimensions 2 and 3.

### Why this is the right cut

This is the first point where the existing pieces become a real linear algebra layer rather than isolated vocabulary:

- `Vector<F, N>` becomes the domain/codomain carrier.
- `OneForm<F, N>` becomes the codomain-1 special case (`F^N -> F`).
- `dot` becomes a bilinear pairing that can be expressed via rows / functionals.
- `IntegerInterval<int>` can later serve as the finite index set abstraction behind the dense carrier, but the MVP does not need to parameterize on intervals yet.

### Explicit non-goals for the MVP

To keep this ticket tractable, I would keep the following out of scope initially:

- determinant / rank / LU / QR / SVD
- sparse storage
- BLAS interop
- eigenvalue algorithms beyond maybe a tiny follow-up power-method example
- generalized tensor algebra
- fully interval-parameterized dependent dimensions (`IntegerInterval` as a type parameter)

### First follow-ups after the MVP

1. **Adjacency matrix as a special linear map**
   - square `LinearMap<bool, N, N>` or `LinearMap<int, N, N>`
   - graph on `IntegerInterval<int>{0, n}`
   - naturally reconnects to the original issue title and to graph algorithms

2. **Covectors / dual space**
   - identify `OneForm<F, N>` as a row-vector / linear functional
   - connect to issue #127

3. **Outer products / rank-1 operators**
   - `u ⊗ v` as `LinearMap<F, M, N>`
   - connects directly to issue #128

### Suggested file/module shape

I would place this in `dedekind.algebra` first, e.g. a new partition such as `:linear_map` or `:matrices`, because the MVP is primarily about algebraic structure and linearity, not yet geometry.

Geometry (`dot`, `norm`, Hilbert structure) can then consume it rather than owning it.

### Acceptance criteria for closing the MVP ticket

- a compile-time-sized dense linear map type exists
- it applies to `Vector<F, N>`
- composition and identity exist for square maps
- tests prove linearity and composition laws
- at least one tiny worked example is present (e.g. 2x2 rotation/shear/scale, or a 3x3 adjacency-style map)

That would be enough to say the repository has bona fide linear algebra support, rather than just vector-space vocabulary.

**Comment by vincentk** [comment]
Follow-up on implementation location: the current MVP should probably live in `dedekind.geometry` first, not `dedekind.algebra`.

Reason: the concrete finite-dimensional carrier we already have is `Vector<F, N>` in `:affine`. A concrete `LinearMap<F, M, N>` / `Matrix<F, M, N>` that applies to `Vector<F, N>` would create a module cycle if introduced directly in `dedekind.algebra`, since `geometry` already imports `algebra`.

So my preferred delivery order is:

1. `LinearMap<F, M, N>` over `Vector<F, N>` in `dedekind.geometry` (honest MVP)
2. Covectors / linear functionals as the `1 x N` case, reconnecting to `OneForm`
3. Outer products / rank-1 operators on top of the same carrier
4. Adjacency-style square maps and graph algorithms as an application layer

If later we introduce a lower-level algebraic vector carrier independent of `geometry::Vector`, the linear-map carrier could be hoisted downward. For the MVP, avoiding the dependency cycle matters more than the perfect namespace placement.

**Comment by vincentk** [comment]
The issue body is empty. Could you describe the intended scope? Specifically:
- What representation should an adjacency matrix use? (Dense matrix over ℕ? Sparse set of edges? A relation R ⊆ V × V?)
- Should this live in the `geometry` module (as a specialisation of the existing matrix/linear-map infrastructure), or in a new `graph` partition of `sets`?
- Is the goal a mathematical object (a functor/relation) or a data structure with efficient lookup?

A one-sentence answer on any of the above would unblock the work.

**Comment by vincentk** [comment]
Captured the scoped set-builder DSL idea (single-operation binding without standalone var declarations) in #156 so it can be implemented independently and non-urgently.

**Comment by vincentk** [comment]
Backlog grooming question (2026-04-16): since #37 is conditional on this issue, could we add a short acceptance-criteria checklist here (expected output/artifacts and a minimal completion definition) to make dependency handoff explicit?

**Comment by vincentk** [comment]
Cross-reference: new epic #319 now tracks the undergraduate graph-theory ontology layer above this adjacency-matrix foundation. In particular, #320 (primitives/incidence), #322 (paths/cycles/distance), and #326 (shortest paths) are direct follow-ons.

**Comment by vincentk** [comment]
Tracking update: epic #330 is active, and child issues #331-#340 are now attached as native subtasks for the algorithms/data-structures ontology-alignment workflow.
